controllers/openstackvmset_controller.go

```diff
@@ -1140,83 +852,770 @@ func (r *OpenStackVMSetReconciler) vmCreateInstance(
 		}
 
 		err := controllerutil.SetControllerReference(instance, vm, r.Scheme)
-
 		if err != nil {
+			cond.Message = fmt.Sprintf("Error set controller reference for %s", vm.Name)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonControllerReferenceError)
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeError)
+			err = common.WrapErrorForObject(cond.Message, instance, err)
+
 			return err
 		}
 
 		return nil
 	})
-
 	if err != nil {
 		return err
 	}
 
 	if op != controllerutil.OperationResultNone {
-		r.Log.Info(fmt.Sprintf("VirtualMachine %s successfully reconciled - operation: %s", vm.Name, string(op)))
+		common.LogForObject(
+			r,
+			fmt.Sprintf("VirtualMachine %s successfully reconciled - operation: %s", vm.Name, string(op)),
+			instance,
+		)
 	}
 
 	hostStatus := instance.Status.VMHosts[ctl.Hostname]
 
 	if vm.Status.Ready {
-		hostStatus.ProvisioningState = ospdirectorv1beta1.VMSetProvisioned
+		hostStatus.ProvisioningState = ospdirectorv1beta1.VMSetCondTypeProvisioned
 	} else if vm.Status.Created {
-		hostStatus.ProvisioningState = ospdirectorv1beta1.VMSetProvisioning
+		hostStatus.ProvisioningState = ospdirectorv1beta1.VMSetCondTypeProvisioning
 	}
 
 	instance.Status.VMHosts[ctl.Hostname] = hostStatus
 
 	return nil
 }
 
-func (r *OpenStackVMSetReconciler) getDeletedVMOSPHostnames(instance *ospdirectorv1beta1.OpenStackVMSet) ([]string, error) {
-
-	var annotatedVMs []string
+//
+// check for DeploymentSSHSecret
+//
+func (r *OpenStackVMSetReconciler) getDeploymentSSHSecret(
+	instance *ospdirectorv1beta1.OpenStackVMSet,
+	cond *ospdirectorv1beta1.Condition,
+) (string, ctrl.Result, error) {
 
-	labelSelector := map[string]string{
+	sshSecret, _, err := common.GetSecret(r, instance.Spec.DeploymentSSHSecret, instance.Namespace)
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			cond.Message = fmt.Sprintf("DeploymentSSHSecret secret does not exist: %v", err)
+			cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonDeploymentSecretMissing)
+			cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeWaiting)
+
+			common.LogForObject(r, cond.Message, instance)
+
+			return "", ctrl.Result{RequeueAfter: 20 * time.Second}, nil
+		}
+		cond.Message = fmt.Sprintf("Error getting DeploymentSSHSecret: %v", err)
+		cond.Reason = ospdirectorv1beta1.ConditionReason(ospdirectorv1beta1.VMSetCondReasonDeploymentSecretError)
+		cond.Type = ospdirectorv1beta1.ConditionType(ospdirectorv1beta1.VMSetCondTypeWaiting)
```

Comment:
Should this maybe be an error condition instead of waiting?  I noticed on line 944 that an error while trying to get the password secret is counted as an error.  Or maybe that one should be the other way around?

Reply:
yes correct, that should be error